include/kernels/gauge_force.cuh

```diff
@@ -5,70 +5,19 @@
 #include <index_helper.cuh>
 #include <thread_array.h>
 #include <kernel.h>
+#include <gauge_path_helper.cuh>
 
 namespace quda {
 
-  struct paths {
-    const int num_paths;
-    const int max_length;
-    int *input_path[4];
-    const int *length;
-    const double *path_coeff;
-    int *buffer;
-    int count;
-
-    paths(int ***input_path, int *length_h, double *path_coeff_h, int num_paths, int max_length) :
-      num_paths(num_paths),
-      max_length(max_length),
-      count(0)
-    {
-      // create path struct in a single allocation
-      size_t bytes = 4 * num_paths * max_length * sizeof(int) + num_paths * sizeof(int);
-      int pad = ((sizeof(double) - bytes % sizeof(double)) % sizeof(double))/sizeof(int);
-      bytes += pad*sizeof(int) + num_paths*sizeof(double);
-
-      buffer = static_cast<int*>(pool_device_malloc(bytes));
-      int *path_h = static_cast<int*>(safe_malloc(bytes));
-      memset(path_h, 0, bytes);
-
-      for (int dir=0; dir<4; dir++) {
-        // flatten the input_path array for copying to the device
-        for (int i=0; i < num_paths; i++) {
-          for (int j=0; j < length_h[i]; j++) {
-            path_h[dir*num_paths*max_length + i*max_length + j] = input_path[dir][i][j];
-            if (dir==0) count++;
-          }
-        }
-      }
-
-      // length array
-      memcpy(path_h + 4 * num_paths * max_length, length_h, num_paths*sizeof(int));
-
-      // path_coeff array
-      memcpy(path_h + 4 * num_paths * max_length + num_paths + pad, path_coeff_h, num_paths*sizeof(double));
-
-      qudaMemcpy(buffer, path_h, bytes, qudaMemcpyHostToDevice);
-      host_free(path_h);
-
-      // finally set the pointers to the correct offsets in the buffer
-      for (int d=0; d < 4; d++) this->input_path[d] = buffer + d*num_paths*max_length;
-      length = buffer + 4*num_paths*max_length;
-      path_coeff = reinterpret_cast<double*>(buffer + 4 * num_paths * max_length + num_paths + pad);
-    }
-    void free() {
-      pool_device_free(buffer);
-    }
-  };
-
-
-  template <typename Float_, int nColor_, QudaReconstructType recon_u, QudaReconstructType recon_m, bool force_>
+  template <typename store_t, int nColor_, QudaReconstructType recon_u, QudaReconstructType recon_m, bool force_>
   struct GaugeForceArg : kernel_param<> {
-    using Float = Float_;
+    using real = typename mapper<store_t>::type;
     static constexpr int nColor = nColor_;
     static constexpr bool compute_force = force_;
+    using Link = Matrix<complex<real>, nColor>;
     static_assert(nColor == 3, "Only nColor=3 enabled at this time");
-    typedef typename gauge_mapper<Float,recon_u>::type Gauge;
-    typedef typename gauge_mapper<Float,recon_m>::type Mom;
+    typedef typename gauge_mapper<real,recon_u>::type Gauge;
```

Comment:
how about using using here for consistency?

Reply:
Fixed locally

Reply:
https://github.com/lattice/quda/pull/1308/commits/ccb7dd9dfe1dbb485f0d2f05456c9276a79fbafb